x/programs/program/memory.go

```diff
@@ -164,31 +167,52 @@ type CallParam struct {
 
 // WriteParams is a helper function that writes the given params to memory if non integer.
 // Supported types include int, uint64 and string.
-func WriteParams(m *Memory, p []CallParam) ([]uint64, error) {
-	params := []uint64{}
+func WriteParams(m *Memory, p []CallParam) ([]SmartPtr, error) {
+	var params []SmartPtr
 	for _, param := range p {
 		switch v := param.Value.(type) {
-		case string:
-			ptr, err := WriteBytes(m, []byte(v))
+		case ids.ID:
+			smartPtr, err := BytesToSmartPtr(v[:], m)
 			if err != nil {
 				return nil, err
 			}
-			params = append(params, uint64(ptr))
-		case int:
-			if v < 0 {
-				return nil, fmt.Errorf("failed to write param: %w", ErrUnderflow)
+			params = append(params, smartPtr)
+		case string:
```

Comment:
why handle strings and ids.ID differently here? Can they just be passed in as SmartPtrs?

Reply:
for now, moving this out of the program module and into simulator, since it isn't used anywhere else.